In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from sklearn.preprocessing import RobustScaler

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
from sklearn.model_selection import cross_val_score, KFold

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import GradientBoostingClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.ensemble import BaggingClassifier

/usr/local/lib/python3.10/dist-packages/dask/dataframe/__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)


In [ ]:
from sklearn.pipeline import Pipeline

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Final players df 2.csv')

Approach 1, on columns extracted from preforming Correlation Matrix

#Preprocessing

In [ ]:
df.head()

result  kills  deaths  assists  damagetochampions        dpm  damageshare  \
0       1      3       1        6              19273   639.5907     0.222049   
1       1      4       1        5              10308   342.0796     0.118761   
2       1      3       2        8              17215   571.2942     0.198339   
3       1      9       0        9              34546  1146.4381     0.398014   
4       1      1       0       17               5454   180.9956     0.062837   

   wardsplaced     wpm  controlwardsbought  ...  opp_xpat10  opp_csat10  \
0           10  0.3319                   6  ...        4125          72   
1           31  1.0288                  12  ...        2983          55   
2           13  0.4314                  13  ...        4766          83   
3           15  0.4978                   4  ...        2503          72   
4           51  1.6925                  14  ...        2431           1   

   golddiffat10  goldat15  xpat15  csat15  opp_goldat15  opp_xpat15  \
0           414      5017    7646     107          4293        6875   
1          -104      5293    5306      93          4955        5309   
2          -114      5333    7356     142          4923        7388   
3           211      5651    5713     132          4477        4479   
4           177      3685    3598       2          3155        3645   

   opp_csat15  golddiffat15  
0         111           724  
1          99           338  
2         132           410  
3         119          1174  
4           5           530  

[5 rows x 34 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108020 entries, 0 to 108019
Data columns (total 34 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   result              108020 non-null  int64  
 1   kills               108020 non-null  int64  
 2   deaths              108020 non-null  int64  
 3   assists             108020 non-null  int64  
 4   damagetochampions   108020 non-null  int64  
 5   dpm                 108020 non-null  float64
 6   damageshare         108020 non-null  float64
 7   wardsplaced         108020 non-null  int64  
 8   wpm                 108020 non-null  float64
 9   controlwardsbought  108020 non-null  int64  
 10  visionscore         108020 non-null  int64  
 11  vspm                108020 non-null  float64
 12  totalgold           108020 non-null  int64  
 13  earnedgold          108020 non-null  int64  
 14  earned gpm          108020 non-null  float64
 15  earnedgoldshare     108020 non-nul

In [ ]:
df.isna().sum()

result                0
kills                 0
deaths                0
assists               0
damagetochampions     0
dpm                   0
damageshare           0
wardsplaced           0
wpm                   0
controlwardsbought    0
visionscore           0
vspm                  0
totalgold             0
earnedgold            0
earned gpm            0
earnedgoldshare       0
goldspent             0
total cs              0
minionkills           0
cspm                  0
goldat10              0
xpat10                0
csat10                0
opp_goldat10          0
opp_xpat10            0
opp_csat10            0
golddiffat10          0
goldat15              0
xpat15                0
csat15                0
opp_goldat15          0
opp_xpat15            0
opp_csat15            0
golddiffat15          0
dtype: int64

#Method

In [ ]:
X = df.drop(columns =['result'])
y = df['result']

Splitting the dataset

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=23)

Gridsearch

In [ ]:
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 10],
    'metric': ['euclidean', 'manhattan', 'chebyshev']
}

In [ ]:
param_grid_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [3,4,5,6]
    }

In [ ]:
param_grid_lr = {
    'C': [0.1, 1, 10, 100],
    'solver': ['liblinear', 'saga']
}

In [ ]:
param_grid_rf = {
    'n_estimators': [10, 100, 200],
    'criterion': ['gini', 'entropy']

}

In [ ]:
param_grid_nb = {
    'var_smoothing': [1e-08, 1e-07, 1e-06, 1e-05]
}

In [ ]:
param_grid_gb = {
    'n_estimators': [10, 100, 200],

    'max_depth': [3, 4, 5, 6]
}

In [ ]:
param_grid_xgb = {
    'n_estimators': [10, 100, 200],

    'max_depth': [3, 4, 5, 6]
}

In [ ]:
param_grid_lgb = {
    'n_estimators': [10, 100, 200],

    'max_depth': [3,4,5,6]
}

In [ ]:
param_grid_bagging = {
    'n_estimators': [10, 100, 200],
    'max_features': [0.5, 0.7, 1.0]
}

Defining the models


In [ ]:
models = {
    'KNN': (KNeighborsClassifier(), param_grid_knn),
    'Decision Tree': (DecisionTreeClassifier(), param_grid_dt),
    'Logistic Regression': (LogisticRegression(max_iter=1000), param_grid_lr),
    'Random Forest': (RandomForestClassifier(), param_grid_rf)

}

In [ ]:
models2 = {

    'Naive Bayes': (GaussianNB(), param_grid_nb),
    'Gradient Boosting': (GradientBoostingClassifier(), param_grid_gb),
    'XGBoost': (XGBClassifier(), param_grid_xgb)

}

In [ ]:
models3 = {

    'LightGBM': (LGBMClassifier(), param_grid_lgb)

}

In [ ]:
models4 = {

    'Bagging': (BaggingClassifier(), param_grid_bagging)

}

using pipeling to scale the models

In [ ]:
for model_name, (model, param_grid) in models.items():


    pipeline = Pipeline([
        ('scaler', RobustScaler()),
        ('model', model)
    ])


    adjusted_param_grid = {f'model__{key}': value for key, value in param_grid.items()}


    grid_search = GridSearchCV(estimator=pipeline, param_grid=adjusted_param_grid, cv=5, scoring='accuracy')
    grid_search.fit(X_train, y_train)


    best_params = grid_search.best_params_
    best_score = grid_search.best_score_
    print(f'{model_name} - Best Params: {best_params}, Best Score: {best_score}')


    best_model = grid_search.best_estimator_
    y_pred = best_model.predict(X_test)
    test_accuracy = accuracy_score(y_test, y_pred)
    print(f'Test Accuracy for {model_name}: {test_accuracy}')

KNN - Best Params: {'model__metric': 'manhattan', 'model__n_neighbors': 10}, Best Score: 0.9050291663039516
Test Accuracy for KNN: 0.9084891686724681
Decision Tree - Best Params: {'model__criterion': 'entropy', 'model__max_depth': 6}, Best Score: 0.8959452017059008
Test Accuracy for Decision Tree: 0.8982132938344751
Logistic Regression - Best Params: {'model__C': 10, 'model__solver': 'saga'}, Best Score: 0.9608174430682721
Test Accuracy for Logistic Regression: 0.9613497500462878
Random Forest - Best Params: {'model__criterion': 'entropy', 'model__n_estimators': 200}, Best Score: 0.9442696009684969
Test Accuracy for Random Forest: 0.9465839659322348


KNN - Best Params: {'model__metric': 'manhattan', 'model__n_neighbors': 10}, Best Score: 0.9050291663039516

Test Accuracy for KNN: 0.9084891686724681

Decision Tree - Best Params: {'model__criterion': 'entropy', 'model__max_depth': 6}, Best Score: 0.8959452017059008

Test Accuracy for Decision Tree: 0.8982132938344751

Logistic Regression - Best Params: {'model__C': 10, 'model__solver': 'saga'}, Best Score: 0.9608174430682721

Test Accuracy for Logistic Regression: 0.9613497500462878

Random Forest - Best Params: {'model__criterion': 'entropy', 'model__n_estimators': 200}, Best Score: 0.9442696009684969

Test Accuracy for Random Forest: 0.9465839659322348

In [ ]:
for model_name, (model, param_grid) in models2.items():


    pipeline = Pipeline([
        ('scaler', RobustScaler()),
        ('model', model)
    ])


    adjusted_param_grid = {f'model__{key}': value for key, value in param_grid.items()}


    grid_search = GridSearchCV(estimator=pipeline, param_grid=adjusted_param_grid, cv=5, scoring='accuracy')
    grid_search.fit(X_train, y_train)


    best_params = grid_search.best_params_
    best_score = grid_search.best_score_
    print(f'{model_name} - Best Params: {best_params}, Best Score: {best_score}')


    best_model = grid_search.best_estimator_
    y_pred = best_model.predict(X_test)
    test_accuracy = accuracy_score(y_test, y_pred)
    print(f'Test Accuracy for {model_name}: {test_accuracy}')

Naive Bayes - Best Params: {'model__var_smoothing': 1e-08}, Best Score: 0.7926193986508516
Test Accuracy for Naive Bayes: 0.7972134789853731
Gradient Boosting - Best Params: {'model__max_depth': 6, 'model__n_estimators': 200}, Best Score: 0.9593015299991994
Test Accuracy for Gradient Boosting: 0.9625995186076652
XGBoost - Best Params: {'model__max_depth': 3, 'model__n_estimators': 200}, Best Score: 0.9609215903145378
Test Accuracy for XGBoost: 0.9621366413627106


Naive Bayes - Best Params: {'model__var_smoothing': 1e-08}, Best Score: 0.7926193986508516

Test Accuracy for Naive Bayes: 0.7972134789853731

Gradient Boosting - Best Params: {'model__max_depth': 6, 'model__n_estimators': 200}, Best Score: 0.9593015299991994

Test Accuracy for Gradient Boosting: 0.9625995186076652

XGBoost - Best Params: {'model__max_depth': 3, 'model__n_estimators': 200}, Best Score: 0.9609215903145378

Test Accuracy for XGBoost: 0.9621366413627106

In [ ]:
for model_name, (model, param_grid) in models3.items():


    pipeline = Pipeline([
        ('scaler', RobustScaler()),
        ('model', model)
    ])


    adjusted_param_grid = {f'model__{key}': value for key, value in param_grid.items()}


    grid_search = GridSearchCV(estimator=pipeline, param_grid=adjusted_param_grid, cv=5, scoring='accuracy')
    grid_search.fit(X_train, y_train)


    best_params = grid_search.best_params_
    best_score = grid_search.best_score_
    print(f'{model_name} - Best Params: {best_params}, Best Score: {best_score}')


    best_model = grid_search.best_estimator_
    y_pred = best_model.predict(X_test)
    test_accuracy = accuracy_score(y_test, y_pred)
    print(f'Test Accuracy for {model_name}: {test_accuracy}')

[LightGBM] [Info] Number of positive: 34563, number of negative: 34569
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.081852 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 6910
[LightGBM] [Info] Number of data points in the train set: 69132, number of used features: 33
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.499957 -> initscore=-0.000174
[LightGBM] [Info] Start training from score -0.000174
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gai

LightGBM - Best Params: {'model__max_depth': 6, 'model__n_estimators': 200}, Best Score: 0.9597528432136635

Test Accuracy for LightGBM: 0.963664136271061


In [ ]:
for model_name, (model, param_grid) in models4.items():

    pipeline = Pipeline([
        ('scaler', RobustScaler()),
        ('model', BaggingClassifier(n_jobs=1))
    ])

    adjusted_param_grid = {f'model__{key}': value for key, value in param_grid.items()}

    grid_search = GridSearchCV(estimator=pipeline, param_grid=adjusted_param_grid, cv=5, scoring='accuracy', n_jobs=-1)
    grid_search.fit(X_train, y_train)

    best_params = grid_search.best_params_
    best_score = grid_search.best_score_
    print(f'{model_name} - Best Params: {best_params}, Best Score: {best_score}')

    best_model = grid_search.best_estimator_
    y_pred = best_model.predict(X_test)
    test_accuracy = accuracy_score(y_test, y_pred)
    print(f'Test Accuracy for {model_name}: {test_accuracy}')

Bagging - Best Params: {'model__max_features': 0.7, 'model__n_estimators': 200}, Best Score: 0.9520806718889328
Test Accuracy for Bagging: 0.955193482688391


Bagging - Best Params: {'model__max_features': 0.7, 'model__n_estimators': 200}, Best Score: 0.9520806718889328

Test Accuracy for Bagging: 0.955193482688391

#Result

KNN - Best Params: {'model__metric': 'manhattan', 'model__n_neighbors': 10}, Best Score: 0.9050291663039516

Test Accuracy for KNN: 0.9084891686724681

Decision Tree - Best Params: {'model__criterion': 'entropy', 'model__max_depth': 6}, Best Score: 0.8959452017059008

Test Accuracy for Decision Tree: 0.8982132938344751

Logistic Regression - Best Params: {'model__C': 10, 'model__solver': 'saga'}, Best Score: 0.9608174430682721

Test Accuracy for Logistic Regression: 0.9613497500462878

Random Forest - Best Params: {'model__criterion': 'entropy', 'model__n_estimators': 200}, Best Score: 0.9442696009684969

Test Accuracy for Random Forest: 0.9465839659322348

Naive Bayes - Best Params: {'model__var_smoothing': 1e-08}, Best Score: 0.7926193986508516

Test Accuracy for Naive Bayes: 0.7972134789853731

Gradient Boosting - Best Params: {'model__max_depth': 6, 'model__n_estimators': 200}, Best Score: 0.9593015299991994

Test Accuracy for Gradient Boosting: 0.9625995186076652

XGBoost - Best Params: {'model__max_depth': 3, 'model__n_estimators': 200}, Best Score: 0.9609215903145378

Test Accuracy for XGBoost: 0.9621366413627106

LightGBM - Best Params: {'model__max_depth': 6, 'model__n_estimators': 200}, Best Score: 0.9597528432136635

Test Accuracy for LightGBM: 0.963664136271061

Bagging - Best Params: {'model__max_features': 0.7, 'model__n_estimators': 200}, Best Score: 0.9520806718889328

Test Accuracy for Bagging: 0.955193482688391